### PR TITLE
ARC-2171 Autocomplete filter reusable form + 3 extra filters

### PR DESCRIPTION
**Pull request by bertyhell**
https://meemoo.atlassian.net/browse/ARC-2171

Make the autocomplete filter form reusable for
* newspaperSeriesName
* locationCreated
* mentions

proxy PR: https://github.com/viaacode/hetarchief-proxy/pull/512

image
image

image
image

image
image

### CODE REVIEW

File: src/modules/visitor-space/components/AutocompleteFieldFilterForm/AutocompleteFieldFilterForm.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+					setForm({ value: '' });
+					clearErrors();
+				},
+				handleSubmit,
+			})}
+		</>
+	);
+};
+
+export default AutocompleteFieldFilterForm;
```
Comment (RobbeBierebeeck) [inline src/modules/visitor-space/components/AutocompleteFieldFilterForm/AutocompleteFieldFilterForm.tsx:105]: I prefer to export the component with a named export directly from the place where it is defined

File: src/modules/visitor-space/components/AutocompleteFieldFilterForm/AutocompleteFieldFilterForm.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import { FormControl } from '@meemoo/react-components';
+import clsx from 'clsx';
+import { isNil } from 'lodash-es';
+import { type FC, type ReactNode, useEffect, useState } from 'react';
+import { Controller, useForm } from 'react-hook-form';
+import { StringParam, useQueryParam } from 'use-query-params';
+import { object as yupObject, string as yupString } from 'yup';
+
+import AutocompleteFieldInput from '@visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput';
+import { AutocompleteField } from '@visitor-space/components/FilterMenu/FilterMenu.types';
```
Comment (RobbeBierebeeck) [inline src/modules/visitor-space/components/AutocompleteFieldFilterForm/AutocompleteFieldFilterForm.tsx:11]: I would shorten the import path a bit, so these two components can come from the same import path. 
`import { AutocompleteFieldInput, AutocompleteField } from '@visitor-space/components'`

File: src/modules/visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput.module.scss [hunk 3]
```diff
@@ -0,0 +1,12 @@
+@use "src/styles/abstracts" as *;
+
+$component: "c-autocomplete-field-input";
+
+.#{$component} {
+	width: 100%;
```
Comment (RobbeBierebeeck) [inline src/modules/visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput.module.scss:6]: I'm not a fan of defining this class in a different component variable. It just makes it more complicated for no reason.

File: src/modules/visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+const AutocompleteFieldInput: FC<AutocompleteFieldInputProps> = ({
+	onChange,
+	value,
+	fieldName,
+	label,
+}) => {
+	const handleLoadOptions = useCallback(
+		(inputValue: string, callback: (newOptions: SelectOption[]) => void): void => {
+			if (inputValue.length < 3) {
```
Comment (RobbeBierebeeck) [inline src/modules/visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput.tsx:30]: Is this 3 we can maybe define in a different const?

File: src/modules/visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	);
+
+	const handleChange = (
+		newValue: SingleValue<SelectOption>,
+		actionMeta: ActionMeta<SelectOption>
+	): void => {
+		if (actionMeta.action === 'select-option' || actionMeta.action === 'create-option') {
+			onChange(newValue?.value || null);
+		}
+	};
```
Comment (RobbeBierebeeck) [inline src/modules/visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput.tsx:48]: Is there a need for this different handleChange component. Can't this be don direcly from the `<AsyncCreatableSelect<SelectOption>` component

File: src/modules/visitor-space/components/DurationFilterForm/DurationFilterForm.tsx [hunk 6]
```diff
@@ -114,7 +114,8 @@ const DurationFilterForm: FC<DurationFilterFormProps> = ({ children, className,
 
 											if (value !== form.operator) {
 												setForm({
-													duration: defaultValues.duration,
+													[SearchFilterId.Duration]:
+														defaultValues.duration,
```
Comment (RobbeBierebeeck) [inline src/modules/visitor-space/components/DurationFilterForm/DurationFilterForm.tsx:118]: It feels a bit inconsistent that the first duration starts with a capital letter and the second duration not

**Review by RobbeBierebeeck — APPROVED** [review]

File: src/modules/visitor-space/components/AutocompleteFieldFilterForm/AutocompleteFieldFilterForm.tsx (hunk 2 above)
Comment (bertyhell) [reply src/modules/visitor-space/components/AutocompleteFieldFilterForm/AutocompleteFieldFilterForm.tsx:11]: that requires a barrel file
and those often have high performance impact on tests
https://marvinh.dev/blog/speeding-up-javascript-ecosystem-part-7/
https://dev.to/tassiofront/barrel-files-and-why-you-should-stop-using-them-now-bc4

File: src/modules/visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput.module.scss (hunk 3 above)
Comment (bertyhell) [reply src/modules/visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput.module.scss]: i agree
but it is the current coding style of the repo
if we decide we want to change it we need to agree with all devs that work on the project before we start changing the code

File: src/modules/visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput.tsx (hunk 5 above)
Comment (bertyhell) [reply src/modules/visitor-space/components/AutocompleteFieldInput/AutocompleteFieldInput.tsx]: you mean: inline in the html code?
i think its cleaner to keep the html code as minimal as possible and make any onChange that isn't just calling the onChange of the props a separate handler in the code.

File: src/modules/visitor-space/components/DurationFilterForm/DurationFilterForm.tsx (hunk 6 above)
Comment (bertyhell) [reply src/modules/visitor-space/components/DurationFilterForm/DurationFilterForm.tsx:118]: one is an enum
the second is a value

There isn't a clear enum casing convention in the code base

Which do you prefer?
I kind of like all caps to signify its a constant

SearchFilterId.DURATION

**Pull request merged**